Track active ssh sessions established via the web

rosstimothy opened the pull request:
Adds open session tracking in `web.Handler` and introduces a new
 `web.Server` to prevent closing the web `http.Server` during
a graceful shutdown if there are still open sessions. From the `http.Server/Shutdown` docs:

> Shutdown does not attempt to close nor wait for hijacked
> connections such as WebSockets. The caller of Shutdown should
> separately notify such long-lived connections of shutdown and wait
> for them to close, if desired.

Because the web sessions use WebSockets, `Shutdown` will not wait for them to be idle before terminating. This used to work due to the fact that web sessions dialed the Proxy SSH Port  to establish the SSH session. Which meant that all accounting of active user connections by `sshutils/server.go` was accurate. But now that sessions are established directly via the `proxy.Router` the web sessions were unaccounted for.

The `web.Server` now mimics the behavior of `Shutdown` in `sshutils/server.go`. Only once all the active web sessions have been terminated will the underlying `http.Server` be shutdown.

Fixes #20227

rosstimothy committed:
Track active ssh sessions established via the web

Adds open session tracking in `web.Handler` and introduces a new
 `web.Server` to prevent closing the web `http.Server` during
a graceful shutdown if there are still open sessions. From the
`http.Server/Shutdown` docs:

> Shutdown does not attempt to close nor wait for hijacked
> connections such as WebSockets. The caller of Shutdown should
> separately notify such long-lived connections of shutdown and wait
> for them to close, if desired.

Because the web sessions use WebSockets, `Shutdown` will not
wait for them to be idle before terminating. This used to work due
to the fact that web sessions dialed the Proxy SSH Port  to establish the
SSH session. Which meant that all accounting of active user connections
by `sshutils/server.go` was accurate. But now that sessions are
established directly via the `proxy.Router` the web sessions were
unaccounted for.

The `web.Server` now mimics the behavior of `sshutils/server.go` for
Shutdown. Only once all the active web sessions have been terminated
will the underlying http.Server be shutdown.

Fixes #20227

zmb3 commented:
FYI @rudream I wonder if this helps with the moderated sessions item we discussed today.

rosstimothy committed:
Refactor web terminal streaming

`web.TerminalStream` replaces `web.terminalStream` and is now
responsible for all reads and writes to the web socket used for
sessions created via the web ui. `web.TerminalHandler` no longer
performs reads/writes directly, instead all operations on the
web socket are now performed via a `web.TerminalStream`. This also
allows for creating and interacting with a web session by other
packages possible.

rosstimothy committed:
Make `intergration.Terminal` an `io.Closer`

rosstimothy committed:
Add helpers for creating a web session

Allows integration tests to create a web session and ssh sessions
for a web session to test interactions from the web UI.

rosstimothy committed:
Update `TestShutdown` to use cli and web sessions

Ensures that graceful shutdown waits for active ssh sessions from
both `tsh` and the web ui to be completed before shutting down the
proxy.

sfreiberg commented on integration/helpers/instance.go:
```diff
@@ -1313,6 +1334,205 @@ func (i *TeleInstance) NewClient(cfg ClientConfig) (*client.TeleportClient, erro
+		return nil, trace.Wrap(err)
+	}
+
+	err = resp.Body.Close()
```
Consider calling this in a defer so that it doesn't leak if any of the errors between dialer.Dial() and this line are hit.

sfreiberg commented on integration/integration_test.go:
```diff
@@ -1329,88 +1331,129 @@ func testShutdown(t *testing.T, suite *integrationTestSuite) {
+			shutdownContext := teleport.Process.StartShutdown(context.Background())
+
+			require.Eventually(t, func() bool {
+				// TODO: check that we either get a connection that fully works or a connection refused error
```
What is this TODO waiting on? Should it be done before this PR is merged?

sfreiberg commented on lib/web/server.go:
```diff
@@ -0,0 +1,140 @@
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// ServerConfig provides dependencies required to create a [Server]
```
Comments should contain punctuation here and below.

sfreiberg commented on lib/web/terminal.go:
```diff
@@ -864,7 +821,141 @@ func serverHostPort(servername string) (string, int, error) {
+// writeError displays an error in the terminal window.
+func (t *TerminalStream) writeError(err error) error {
+	// Replace \n with \r\n so the message correctly aligned.
+	r := strings.NewReplacer("\n", "\r\n")
```
Is it a concern that this would replace "\r\n" with "\r\r\n"? Also since it's only used once why not replace this with strings.ReplaceAll?

rosstimothy committed:
fix: add punctuation to comments

rosstimothy committed:
fix: reuse string replacer

rosstimothy committed:
fix: defer response body closure

timothyb89 reviewed (approved):
Sorry for the delay, this seems good!
